### ISSUE TITLE
[v1.13] Connection disruption on agent restart due to XfrmInStateProtoError errors when IPSec per-tunnel keys are enabled 

### ISSUE THREAD

**Issue by giorio94** [cilium/cilium#33881]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Version

higher than v1.13.18 and lower than v1.14.0

### What happened?

Discovered while attempting to backport https://github.com/cilium/cilium/pull/33769 to v1.13 and v1.14. The Cilium IPsec upgrade tests failed due to connection disruption and increase of `XfrmInStateProtoError` errors, and more specifically:

* On the v1.14 branch, when downgrading from v1.14 (tip) to v1.13.18 (e.g., https://github.com/cilium/cilium/actions/runs/9939762286/job/27454998914)
* On the v1.13 branch, when upgrading from v1.13.18 to v1.13 (tip) (e.g., https://github.com/cilium/cilium/actions/runs/9943013136/job/27465833835)

Given the latter failure, and the fact that the tip of v1.13 practically matches v1.13.18, this issues is most likely happening on agent restart.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by jschwinger233** [comment]
Not very resultful, some interesting facts:
1. XfrmInStateProtoError isn't transient, I can see it keep going up after upgrade completes.
2. It's a general breakdown, connectivity is gone even across new pods.
3. `ip x s` on all nodes looks good I think? SPIs and aead keys are all right. 
4. Dropped packets are carrying correct SPI.

Suspicious points:
1. `ip x s` shows last used time like `lastused 2024-07-18 18:33:12`, I see some of these were last used long time ago. This may imply the kernel did xfrm lookup and matched a wrong sa for decryption?

Actions I'll do tomorrow:
1. I can use bpftrace to see what xfrm policy/state is found for encryption and decryption, to see if they match my expectation.
2. I can compare the `ip x s` and `ip x p` with the original xfrm configures, to see if there is anything leftover / missing / changed slightly.
3. I can tcpdump the dropped esp packets and try to decrypt them in wireshark, to confirm the encryption key is what we expect.

**Comment by julianwiedmann** [comment]
> Given the latter failure, and the fact that the tip of v1.13 practically matches v1.13.18, this issues is most likely happening on agent restart.

If this holds true and agent restart is sufficient to repro, it might be worth to roughly bisect the culprit (or see whether this affects even the v1.13 patch release that introduced support for the per-node-pair key system).

**Comment by jschwinger233** [comment]
Thank you Julian, but restarting agent couldn't reproduce it 100%, so I didn't git bisect. Thank you all the same!

I haven't made good progress, will continue.

The direct reason is mismatching anti-play seq/oseq. 

Consider pod1@host1 -> pod2@host2 via ipsec, the src host1 has an xfrm state:

```
src 10.244.1.177 dst 10.244.2.223
	proto esp spi 0x00000003 reqid 1 mode tunnel
	replay-window 0 flag esn
	mark 0x6ec50d00/0xffff0f00 output-mark 0xd00/0xffffff00
	aead rfc4106(gcm(aes)) 0xbe0f550e9e3282395cec9b4efaee868fd056879f 128
	lastused 2024-07-19 22:14:18
	anti-replay esn context:
	seq-hi 0x0, seq 0x2bb8d7, oseq-hi 0x0, oseq 0x0
	replay_window 1024, bitmap-length 32
	ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff 
	ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff 
	ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff 
	ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff 
	sel src 0.0.0.0/0 dst 0.0.0.0/0 
```

and dst host2 has an xfrm state:

```
src 10.244.1.177 dst 10.244.2.223
	proto esp spi 0x00000003 reqid 1 mode tunnel
	replay-window 0 flag esn
	mark 0xa5a03e00/0xffffff00 output-mark 0xe00/0xffffff00
	aead rfc4106(gcm(aes)) 0xbe0f550e9e3282395cec9b4efaee868fd056879f 128
	lastused 2024-07-19 22:14:18
	anti-replay esn context:
	seq-hi 0x0, seq 0x0, oseq-hi 0x0, oseq 0x2bb8d7
	replay_window 1024, bitmap-length 32
	00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 
	00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 
	00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 
	00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 
	sel src 0.0.0.0/0 dst 0.0.0.0/0 
```

The key to this issue is the replay-seq at src host and replay-oseq at dst host. The above xfrm states are matching each other, src host has replay-seq 0x2bb8d7 and dst host has reply-oseq 0x2bb8d7, two number must be same.

What happened during upgrade is, node2 agent restarted and decided to replace xfrm states (by removing current ones and adding new ones). Unfortunately our cilium-agent always creates xfrm state with zero replay-seq/oseq, but the src host is still encrypting packets using non-zero replay-seq, that leads to dst host drops packets due to replay-seq mismatch.

Why we only see issue on 1.14 -> 1.13 downgrade? My guess is agent has a logic to check whether host already has the needed xfrm states, and do nothing if there are. For higher version agent doesn't replace xfrm states and we luckily never saw the issue. For 1.13, maybe something slightly different in code makes agent replace xfrm states, flush replay-seq/oseq.

(to be continued)

**Comment by pchaigno** [comment]
(I wasn't aware this could also cause XfrmInStateProtoError. We will need to update the IPsec troubleshooting guide to mention that.)

**Comment by pchaigno** [comment]
cc @rgo3 might be good for you to read this thread since it looks somewhat related to the BootID investigation you conducted some time ago.

**Comment by giorio94** [comment]
> This is the part I still don't understand: cilium config -a | grep -i ipam says it's kubernetes, but 1.13 agent says it's cluster-pool (by adding log). So maybe something in ipam mode settings makes a difference? That's everything I have for now, will continue tomorrow.

Beware of the following subtlety with respect to Cilium CLI and IPAM mode, which already confused me in the past. 

TL;DR; Although the default IPAM mode [configured in the values file](https://github.com/cilium/cilium/blob/b0a69f3cbcb1b05bf3eca677b4cabee55b260ec4/install/kubernetes/cilium/values.yaml#L1773) is `cluster-pool`, the [CIlium CLI configures](https://github.com/cilium/cilium-cli/blob/131ed220dc5d2f8358d9a3aa5f587f589cd5da87/install/helm.go#L33-L35) the `kubernetes` IPAM mode for kind cluster. However, this setting takes only effect for the initial installation, and may get reset (depending on the additional flags) during subsequent upgrade/downgrade operations. Hence, you may end up with a cluster initially running in `kubernetes` IPAM mode transitioning into `cluster-pool` mode. 

Although in most cases this does not lead to actual issues on Kind (as the PodCIDR field of CiliumNodes is not subsequently modified), it may definitely trigger different code paths. If you observe this issue, you may want to explicitly configure the IPAM mode to prevent these types of inconsistencies, as done for instance [here](https://github.com/cilium/cilium/commit/7d2505ee18a0015df500de8a4618f94452e1d034) for the clustermesh upgrade/downgrade workflow.

**Comment by rgo3** [comment]
Not contributing to the root-cause investigation but I do recall that we had the following patch (which we never merged) that would ensure that we wouldn't update xfrm states if the triggering node update contained an empty bootid: 
```diff
diff --git a/pkg/datapath/linux/ipsec.go b/pkg/datapath/linux/ipsec.go
index 27d2491aef..e85d409498 100644
--- a/pkg/datapath/linux/ipsec.go
+++ b/pkg/datapath/linux/ipsec.go
@@ -84,7 +84,7 @@ func (n *linuxNodeHandler) enableIPsec(oldNode, newNode *nodeTypes.Node, nodeID
 		}
 	}

-	if oldNode != nil && oldNode.BootID != newNode.BootID {
+	if oldNode != nil && oldNode.BootID != newNode.BootID && newNode.BootID != "" {
 		n.ipsecUpdateNeeded[newNode.Identity()] = true
 	}
 	_, updateExisting := n.ipsecUpdateNeeded[newNode.Identity()]
 ```
 We never merged it because it wasn't clear if it was strictly necessary to include it, and the manual testing of it wasn't conclusive. But after reading through grays investigation it seems like it could have prevented the issue?

**Comment by julianwiedmann** [comment]
Fixed by https://github.com/cilium/cilium/pull/34035. Thank you all!

**Comment by j-land** [comment]
I'm pretty sure we are seeing the same root issue in 1.15.7 - every restart of the cilium daemonset results in `XfrmInStateProtoError` errors in all cilium agent pods. Could you backport this fix to 1.15?

**Comment by julianwiedmann** [comment]
> I'm pretty sure we are seeing the same root issue in 1.15.7 - every restart of the cilium daemonset results in `XfrmInStateProtoError` errors in all cilium agent pods. Could you backport this fix to 1.15?

If I remember correctly the fix was not applicable to v1.15 - so please open a fresh issue with the relevant info, you're most likely seeing a different problem.
